File: AllInOne/src/Greedy/GasStation.java
package Greedy;

/**
 * 
 * @author jingjiejiang created on Sep 5, 2018
 * @ref: https://leetcode.com/problems/gas-station/discuss/164669/Java-Clean-O(n)-and-constant-space-solution-beats-100
 */
public class GasStation {

	// from brutal force to O(n) solution
	public static int canCompleteCircuit(int[] gas, int[] cost) {
		
        int count = 0;
        int pos = 0;
        int sum = 0;
        
        for (int i = 0; i < gas.length; i ++) {
        	count += gas[i] - cost[i]; // temporary sum
        	sum += gas[i] - cost[i];
        	
        	if (count < 0) {
        		count = 0;
        		pos = (i + 1) % gas.length;
        	}
        }
        
        return sum >= 0? pos : -1;
    }

	public static void main(String[] args) {
		
		int[] gas = new int[]{1,2,3,4,5};
		int[] cost = new int[]{3,4,5,1,2};
		System.out.println(canCompleteCircuit(gas, cost));
	}
}

File: AllInOne/src/Greedy/GasStation.ts
/**
 * 134. Gas Station
 */
function canCompleteCircuit(gas: number[], cost: number[]): number {
  // binary search where left < mid and mid > right
  let curGas = 0,
    totalGas = 0,
    pos = 0;

  if (!gas || !cost || gas.length !== cost.length) return undefined;

  for (let idx = 0; idx < gas.length; idx++) {
    curGas += gas[idx] - cost[idx];
    totalGas += gas[idx] - cost[idx];

    if (curGas < 0) {
      pos = idx + 1;
      curGas = 0;
    }
  }

  return totalGas >= 0 ? pos : -1;
}
